Add support for custom actions

SanjulaGanepola opened the pull request:
### Changes

With the removal of `iproj.json` actions in PR #1481, this PR adds the ability to run custom actions with custom deployment method. Also, updates `launchActionSetup` to accept a workspace folder.

### Checklist

* [x] have tested my change
* [ ] updated relevant documentation
* [x] Remove any/all `console.log`s I added
* [x] eslint is not complaining
* [ ] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/halcyon-tech/vscode-ibmi/blob/master/CONTRIBUTING.md)
* [x] **for feature PRs**: PR only includes one feature enhancement.

worksofliam commented:
@halcyon-tech/core I need some additional eyes on this PR. I am happy with the code and have reviewed with Sanjula, but I'd like a second opinion. I'd also like some additional testing on this for different areas where Actions can be run.

* Running an action from active editors
   * Members
   * IFS
   * Local
* Running actions from views
   * Object Browser
   * IFS Browser

Thank you!

sebjulliand commented on src/instantiate.ts:
```diff
@@ -214,7 +214,7 @@ export async function loadAllofExtension(context: vscode.ExtensionContext) {
       quickPick.show();
     }),
-    vscode.commands.registerCommand(`code-for-ibmi.runAction`, async (node) => {
+    vscode.commands.registerCommand(`code-for-ibmi.runAction`, async (node: any, group?: any, action?: Action, method?: DeploymentMethod) => {
```
`group` seems to be unused here.
```suggestion
    vscode.commands.registerCommand(`code-for-ibmi.runAction`, async (node: any, action?: Action, method?: DeploymentMethod) => {
```

SanjulaGanepola replied:
Although this parameter is unused, we still need it because when `code-for-ibmi.runAction` is invoked from the editor's context menu, VS Code passes in this parameter (it has to do with the editor tab group).

![image](https://github.com/halcyon-tech/vscode-ibmi/assets/32170854/089cbaa4-e9a1-48ae-995c-13ef005dc682)

sebjulliand replied:
Ha right, good catch 😅

sebjulliand commented on src/instantiate.ts:
```diff
@@ -245,7 +245,7 @@ export async function loadAllofExtension(context: vscode.ExtensionContext) {
 
           if (canRun && [`member`, `streamfile`, `file`].includes(uri.scheme)) {
-            CompileTools.runAction(instance, uri);
+            CompileTools.runAction(instance, uri, action, method);
```
We should `await` here if the caller of the command want to actually wait for it to end.
```suggestion
            await CompileTools.runAction(instance, uri, action, method);
```

SanjulaGanepola replied:
Good point. Update that: cf3a4a5

sebjulliand reviewed (changes requested):
Code looks good. I just suggested minor changes.

I tested the following cases:
- Run action from current editor (IFS, Member and Local)
- Run action from IFS and Object browsers

Every action ran as expected.

A few other changes to consider though:
- Maybe the `extensions` field in `Action` should be optional now.
- Add a test for the `code-for-ibmi.runAction`, to be on the safe side (with an `Action` defined on the fly).

SanjulaGanepola commented:
I made the following changes to the tests:
- Instead of testing `CompileTools.runAction` directly, I updated the tests to call the `code-for-ibmi.runAction` command as this gives us better coverage
- Added 2 additional tests for custom actions against members and streamfiles

![image](https://github.com/halcyon-tech/vscode-ibmi/assets/32170854/3f6f9f57-ba5a-4f0c-b152-9338bf4108f4)

SanjulaGanepola commented:
@sebjulliand As you suggested, I updated the `Actions` interface so that the `extensions` property is now optional -> [98a99f8](https://github.com/halcyon-tech/vscode-ibmi/pull/1484/commits/98a99f8392207e4db03ba0d5066f0ee260d1958c)

However, the actions.json schema was not changed.

worksofliam reviewed (approved):
I have tested the following scenarios:

* Running user defined actions for members
* Running user defined actions for local files
* Test cases are passing

I'd also like an additional approval from someone in @halcyon-tech/core before this gets merged. Thanks!

sebjulliand reviewed (approved):
Looks good and work as expected!
